Add info for contributors in WASM readme

zploskey opened the pull request:
This expands the README for the WASM backend to include some of the information we discussed in the hangout earlier today. It is far from a complete description, but attempts to provide an overview of this part of the codebase. No doubt this could be expanded further. Please let me know if anything needs to be corrected.

zploskey committed:
Add info for contributors in WASM readme

zploskey commented:
I would also like to add some instructions for getting started. That may currently be blocked by a few things, but perhaps we could start with what dependencies need to be installed. In trying to do the normal build process with `./configure` then `make world.opt`, I am getting this:
```sh
...
gcc -c -O2 -fno-strict-aliasing -fwrapv -Wall -Werror -fno-tree-vrp  -g -DDEBUG -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DCAML_NAME_SPACE  -o instrtrace.d.o instrtrace.c
rm -f libcamlruni.a && emar rc libcamlruni.a  interp.i.o misc.i.o stacks.i.o fix_code.i.o startup_aux.i.o startup.i.o freelist.i.o major_gc.i.o minor_gc.i.o memory.i.o alloc.i.o roots.i.o globroots.i.o fail.i.o signals.i.o signals_byt.i.o printexc.i.o backtrace_prim.i.o backtrace.i.o compare.i.o ints.i.o floats.i.o str.i.o array.i.o io.i.o extern.i.o intern.i.o hash.i.o sys.i.o meta.i.o parsing.i.o gc_ctrl.i.o md5.i.o obj.i.o lexing.i.o callback.i.o debugger.i.o weak.i.o compact.i.o finalise.i.o custom.i.o dynlink.i.o spacetime.i.o afl.i.o unix.i.o bigarray.i.o main.i.o; emranlib libcamlruni.a libcamlruni.a
(echo '#define CAML_INTERNALS'; \
         echo '#include "caml/mlvalues.h"'; \
 echo '#include "caml/prims.h"'; \
 sed -e 's/.*/extern value &();/' primitives; \
 echo 'c_primitive caml_builtin_cprim[] = {'; \
 sed -e 's/.*/	&,/' primitives; \
 echo '	 0 };'; \
 echo 'char * caml_names_of_builtin_cprim[] = {'; \
 sed -e 's/.*/	"&",/' primitives; \
 echo '	 0 };') > prims.c
/bin/sh: 1: emar: not found
/bin/sh: 1: emranlib: not found
Makefile:200: recipe for target 'libcamlruni.a' failed
make[1]: *** [libcamlruni.a] Error 127
make[1]: *** Waiting for unfinished jobs....
make[1]: Leaving directory '/home/zach/src/ocaml-wasm/byterun'
Makefile:413: recipe for target 'coldstart' failed
make: *** [coldstart] Error 2
```
So we at least need `emscripten` available (not sure on version requirements), and apparently a fork of LLVM.

exced commented:
I ran into the same problem trying to execute `make world.opt`. I'm trying to install dependencies and will create a script or Dockerfile afterwards

zploskey commented:
I'm on Ubuntu 18.04, and I tried to proceed by just installing the `emscripten` package provided by the system. It proceeded further after that. Please note that you may need to run the build twice for it to work if it's the first time you've run emscripten commands, as it seems to stop with a warning about where it thinks the paths of things are. I then get to this error:

```sh
...
gcc -O2 -fno-strict-aliasing -fwrapv -Wall -Werror -fno-tree-vrp -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DCAML_NAME_SPACE  -Wl,-E -o ocamlyacc closure.o error.o lalr.o lr0.o main.o mkpar.o output.o reader.o skeleton.o symtab.o verbose.o warshall.o 
make[1]: Leaving directory '/home/zach/src/ocaml-wasm/yacc'
cp yacc/ocamlyacc boot/ocamlyacc
make -C stdlib  \
  COMPILER="../boot/ocamlc -use-prims ../byterun/primitives" all
make[1]: Entering directory '/home/zach/src/ocaml-wasm/stdlib'
../boot/ocamlrun ../boot/ocamlc -use-prims ../byterun/primitives -strict-sequence -absname -w +a-4-9-41-42-44-45-48 -g -warn-error A -bin-annot -dcmm -nostdlib -safe-string -strict-formats  -nopervasives -c camlinternalFormatBasics.mli
sed -e "s|%%VERSION%%|`sed -e 1q ../VERSION | tr -d '\r'`|" sys.mlp > sys.ml
for suff in '' d i; do \
  echo '#!/usr/local/bin/ocamlrun'$suff > camlheader$suff && \
  echo '#!/usr/local/bin/ocamlrun'$suff >target_camlheader$suff; \
done && \
echo '#!' | tr -d '\012' > camlheader_ur;
for suff in '' d i; do \
  echo '#!/usr/local/bin/ocamlrun'$suff > camlheader$suff && \
  echo '#!/usr/local/bin/ocamlrun'$suff >target_camlheader$suff; \
done && \
echo '#!' | tr -d '\012' > camlheader_ur;
for suff in '' d i; do \
  echo '#!/usr/local/bin/ocamlrun'$suff > camlheader$suff && \
  echo '#!/usr/local/bin/ocamlrun'$suff >target_camlheader$suff; \
done && \
echo '#!' | tr -d '\012' > camlheader_ur;
for suff in '' d i; do \
  echo '#!/usr/local/bin/ocamlrun'$suff > camlheader$suff && \
  echo '#!/usr/local/bin/ocamlrun'$suff >target_camlheader$suff; \
done && \
echo '#!' | tr -d '\012' > camlheader_ur;
for suff in '' d i; do \
  echo '#!/usr/local/bin/ocamlrun'$suff > camlheader$suff && \
  echo '#!/usr/local/bin/ocamlrun'$suff >target_camlheader$suff; \
done && \
echo '#!' | tr -d '\012' > camlheader_ur;
../boot/ocamlc: unknown option '-dcmm'.
Usage: ocamlc <options> <files>
<... snipped usage information ...>
Makefile:234: recipe for target 'camlinternalFormatBasics.cmi' failed
make[1]: *** [camlinternalFormatBasics.cmi] Error 2
make[1]: Leaving directory '/home/zach/src/ocaml-wasm/stdlib'
Makefile:413: recipe for target 'coldstart' failed
make: *** [coldstart] Error 2
```
The bootstrapped `ocamlc` claims not to have a `-dcmm` option, and I'm not sure what to do about that. Maybe there's something missing at the configure step that's needed for that? Maybe cmm needs to be installed before configuring.

SanderSpies commented:
If you want to build the system manually:
- remove the `-dcmm` option in `stdlib/Makefile`. I use it for debugging purposes, but it should be removed when initially building everything.
- install llvm from source via these instructions: `https://gist.github.com/yurydelendik/4eeff8248aeb14ce763e`
- replace the hardcoded references to lld in `utils/ccomp` with the right ones on your machine
- run `make wasm32-all` to have an initial setup
- from there you can run `make wasm32` to iterate on wasm
there will most likely be a few errors due to a missing test file, but nothing shocking. If you see `THIS WORKS` it's good. 

Note that I'm trying to get this working inside docker.

SanderSpies commented:
btw. emscripten is not required, I don't use it all - I did do an experiment with it a while ago and need to remove the left overs.

SanderSpies commented on asmcomp/wasm32/README.md:
```diff
@@ -8,5 +8,52 @@ cmm -> typed cmm -> wasm ast -> linking -> binary
 ---
 
+In normal OCaml compilation, after typechecking has been completed the types are removed,
+and thus the cmm representation does not include types.
```
but only memory representations.

zploskey replied:
So if I understood correctly, the typed_cmm extracted from the encoded memory representations in the cmm is used to annotate the necessary types in the WASM AST that is produced next. Basically it extracts the type info necessary to convert to wasm equivalents for floats, ints, vals and memory addresses, etc.

SanderSpies commented on asmcomp/wasm32/README.md:
```diff
@@ -8,5 +8,52 @@ cmm -> typed cmm -> wasm ast -> linking -> binary
 
+In normal OCaml compilation, after typechecking has been completed the types are removed,
+and thus the cmm representation does not include types.
+However, for WASM, adding some types back at this stage is helpful to cleanly generate the WebAssembly,
```
This is phrased a bit wrong, so what happens is that the types are removed - and you do not get them back - instead it's now limited to `int`'s and `float`s as described here: `https://github.com/SanderSpies/ocaml/blob/wasm-backend/asmcomp/cmm.mli#L18`. OCaml can guarantee that everything is correct after type checking and therefore does not need to check if the type is correct. WASM can't make this assumption and therefore does require extra type information. With typed_cmm extra type information is added where webassembly also requires it. I recommend reading https://v1.realworldocaml.org/v1/en/html/memory-representation-of-values.html for more information on this.

SanderSpies commented on asmcomp/wasm32/README.md:
```diff
@@ -8,5 +8,52 @@ cmm -> typed cmm -> wasm ast -> linking -> binary
+and thus the cmm representation does not include types.
+However, for WASM, adding some types back at this stage is helpful to cleanly generate the WebAssembly,
+because CMM is closer to WASM than it is to the typed AST.
+WebAssembly types and helper functions are defined in [typed_cmm.ml](typed_cmm.ml).
```
Incorrect, `typed_cmm` only enhances the cmm provided by OCaml. Within `emit.mlp` the `typed_cmm` is used to translate it to the webassembly ast, which then gets transformed in actual binary code within `encode.ml`.

SanderSpies commented on asmcomp/wasm32/README.md:
```diff
@@ -8,5 +8,52 @@ cmm -> typed cmm -> wasm ast -> linking -> binary
 Linking
 ---
+
+We currently depend on a fork of LLVM for linking with other WASM code.
```
The fork is only required for wasm gc, we want to upstream the work at some point.

SanderSpies commented on asmcomp/wasm32/README.md:
```diff
@@ -8,5 +8,52 @@ cmm -> typed cmm -> wasm ast -> linking -> binary
+---
+
+The entry-point for WASM compilation is [asmcomp/asmgen.mlp-wasm](../asmgen.mlp-wasm).
+The main points of interest in the pipline are:
```
pipeline

zploskey committed:
Address comments

zploskey commented:
@SanderSpies I think that addresses all the comments so far. Let me know if you have any other tweaks or suggestions. I guess we could add the build instructions as you described them for those that don't want to rely on docker.

SanderSpies reviewed (approved):
Thanks.
